icommands.c

```diff
@@ -98,6 +103,173 @@ enum CommandResult mutt_parse_icommand(/* const */ char *line, struct Buffer *er
+  if (map->desc)
+  {
+    fprintf(fpout, "macro %s %s \"%s\" \"%s\"\n", menu_name, key_binding,
+            tmp.data, map->desc);
+  }
+  else
+  {
+    fprintf(fpout, "macro %s %s \"%s\"\n", menu_name, key_binding, tmp.data);
+  }
+}
+
+/**
+ * dump_bind - Dump a bind to a file
+ * @param fpout     Output file
+ * @param menu      Menu id of the bind
+ * @param menu_name Name of the bind's menu
+ * @param map       Bind keymap
+ */
+static void dump_bind(FILE *fpout, const int menu, const char *menu_name, struct Keymap *map)
+{
+  char key_binding[MAX_SEQ];
+  const char *fn_name = NULL;
+
+  km_expand_key(key_binding, MAX_SEQ, map);
+  if (map->op == OP_NULL)
+  {
+    fprintf(fpout, "bind %s %s noop\n", menu_name, key_binding);
+    return;
+  }
+
+  /* The pager and editor menus don't use the generic map,
+   * however for other menus try generic first. */
+  if ((menu != MENU_PAGER) && (menu != MENU_EDITOR) && (menu != MENU_GENERIC))
+  {
+    fn_name = mutt_get_func(OpGeneric, map->op);
+  }
+
+  /* if it's one of the menus above or generic doesn't find
+   * the function, try with it's own menu. */
+  if (!fn_name)
+  {
+    const struct Binding *bindings = km_get_table(menu);
+    if (!bindings)
+      return;
+
+    fn_name = mutt_get_func(bindings, map->op);
+  }
+
+  fprintf(fpout, "bind %s %s %s\n", menu_name, key_binding, fn_name);
+}
+
+/**
+ * bind_dump_menu - Dumps all the binds/macros of a menu into a file.
+ * @param fpout     Output file
+ * @param menu      Menu id to dump
+ * @param menu_name Name of the menu to dump
+ * @param bind      If true it's :bind icommand, else :macro
+ */
+static void dump_menu(FILE *fpout, const int menu, const char *menu_name, bool bind)
+{
+  struct Keymap *map = NULL, *next = NULL;
+
+  for (map = Keymaps[menu]; map; map = next)
+  {
+    next = map->next;
+
+    if (!bind && map->op == OP_MACRO)
+      dump_macro(fpout, menu_name, map);
+    else if (bind && map->op != OP_MACRO)
+      dump_bind(fpout, menu, menu_name, map);
+  }
+}
+
+/**
+ * icmd_bind - Parse 'bind' and 'macro' commands - Implements ::icommand_t
+ */
+static enum CommandResult icmd_bind(struct Buffer *buf, struct Buffer *s,
+                                    unsigned long data, struct Buffer *err)
+{
+  FILE *fpout = NULL;
+  bool valid_menu = false;
+  char tempfile[PATH_MAX];
+
+  if (!MoreArgs(s))
+    mutt_buffer_strcpy(buf, "all");
+  else
+    mutt_extract_token(buf, s, 0);
+
+  if (MoreArgs(s))
+  {
+    /* More arguments potentially means the user is using the
+     * ::command_t :bind command thus we delegate the task. */
+    return MUTT_CMD_ERROR;
+  }
+
+  mutt_mktemp(tempfile, sizeof(tempfile));
+  fpout = mutt_file_fopen(tempfile, "w");
+  if (!fpout)
+  {
+    mutt_buffer_printf(err, _("Could not create temporary file %s"), tempfile);
+    return MUTT_CMD_ERROR;
+  }
+
+  const bool bind = data == 0;
+  const bool dump_all = mutt_str_strcasecmp(buf->data, "all") == 0;
+  for (int i = 0; i < MENU_MAX; i++)
```

Comment:
I feel like this `for` loop exists for the sake of `dump_all`. If we're not dumping everything, we can use `mutt_map_get_value()` instead of looking for an appropriate menu.

Reply:
If we didn't worry about `dump_all` support, would the code be simplified to the following?

```C
const int menu_index = mutt_mapping_get_value(...);

// handle no menu error

dump_menu(...);
```

If so, I think we should write the code as:

```C
if (dump_all)
    dump_all_menus(...); // contains our for loops
else
{
    const int menu_index = mutt_mapping_get_value(...);

    // handle no menu error

    dump_menu(...);
}
```

Then the code would have less mixture between low-level and high-level policy, no special special cases in the `for` loop, and we can unit test the `dump_all_menus(...)` method!

Reply:
That's a great improvement to the code quality! I applied these suggestions already